posthog/tasks/process_event.py

```diff
@@ -1,32 +1,67 @@
+    old_person: Optional[Person] = None
+    new_person: Optional[Person] = None
+
     try:
-        person = Person.objects.get(team_id=team_id, persondistinctid__distinct_id=distinct_id)
+        old_person = Person.objects.get(team_id=team_id, persondistinctid__distinct_id=previous_distinct_id)
     except Person.DoesNotExist:
-        # no reason to alias if it doesn't exist
-        # mostly happens if someone calls identify before the user has done anything
-        return
+        pass
+
     try:
-        person.add_distinct_id(new_distinct_id)
-    except IntegrityError:
-        # IntegrityError means a person with new_distinct_id already exists
-        # That can either mean `person` already has that distinct_id, in which case we do nothing
-        # OR it means there is _another_ Person with that distinct _id, in which case we want to remove that person
-        # and add that distinct ID to `person`
-        previous_person = Person.objects.filter(team_id=team_id, persondistinctid__distinct_id=new_distinct_id).exclude(pk=person.id)
-        if previous_person.exists():
-            person.properties.update(previous_person.first().properties) # type: ignore
-            previous_person.delete()
-            person.add_distinct_id(new_distinct_id)
-            person.save()
+        new_person = Person.objects.get(team_id=team_id, persondistinctid__distinct_id=distinct_id)
+    except Person.DoesNotExist:
+        pass
+
+    if old_person and not new_person:
+        try:
+            old_person.add_distinct_id(distinct_id)
+        # Catch race case when somebody already added this distinct_id between .get and .add_distinct_id
+        except IntegrityError:
+            if retry_if_failed:  # run everything again to merge the users if needed
+                _alias(previous_distinct_id, distinct_id, team_id, False)
+        return
+
+    if not old_person and new_person:
+        try:
+            new_person.add_distinct_id(previous_distinct_id)
+        # Catch race case when somebody already added this distinct_id between .get and .add_distinct_id
+        except IntegrityError:
+            if retry_if_failed:  # run everything again to merge the users if needed
+                _alias(previous_distinct_id, distinct_id, team_id, False)
+        return
+
+    if not old_person and not new_person:
+        try:
+            Person.objects.create(team_id=team_id, distinct_ids=[str(distinct_id), str(previous_distinct_id)])
+        # Catch race condition where in between getting and creating, another request already created this user.
+        except IntegrityError:
+            if retry_if_failed:
+                # try once more, probably one of the two persons exists now
+                _alias(previous_distinct_id, distinct_id, team_id, False)
+        return
+
+    if old_person and new_person:
+        if old_person == new_person:
+            return
+
+        new_person.properties.update(old_person.properties)
```

Comment:
There's one issue here I think.
What if `old_person` has a property "plan" = "free" and `new_person` has "plan" = "premium"
Won't it override "premium" to "free" then?
I'll check this. Please don't merge until then :)

Reply:
Fixed with the [last commit](https://github.com/PostHog/posthog/pull/637/commits/50078986a679317c25475f83905dbf6841762e28). Now the properties are merged, yet the properties of `new_person` take precedence.

Reply:
Great find